### PR TITLE
base module: `NullUtils` -> `internal`

### PR DESCRIPTION
**Pull request by RingerJK**
## Description

`NullUtils` duplicates on `libnavigation-base` and `libnavigation-util` because `libnavigation-base` doesn't depend on `libnavigation-util`. Other modules should use `NullUtils` from `libnavigation-util`

- [ ] I have added any issue links
- [x] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [x] I have added the appropriate milestone and project boards

### Goal

Please describe the PR goals. Just the stuff needed to implement the fix / feature and a simple rationale. It could contain many check points if needed

### Implementation

Please include all the relevant things implemented and also rationale, clarifications / disclaimers etc. related to the approach used. It could be as self code companion comments

## Screenshots or Gifs

Please include all the media files to give some context about what's being implemented or fixed. It's not mandatory to upload screenshots or gifs, but for most of the cases it becomes really handy to get into the scope of the feature / bug being fixed and also it's REALLY useful for UI related PRs screenshot gif

## Testing

Please describe the manual tests that you ran to verify your changes

- [ ] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [ ] I have tested via a test drive, or a simulation/mock location app
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR

### CODE REVIEW

**Comment by evabishchevich** [comment]
@RingerJK Is it worth to have this utils duplicated? 
We use `base` module inside all others, so I'll vote to have this utils only in `base` module despite the fact that it's utils

**Comment by RingerJK** [comment]
@evabishchevich great question, I see 2 points why we should have these changes:
- base module provided transitively(`utils` is provided internally) and everything that we have public is available for customers. As less we have open as less we should support 👍 
- unit module is more suitable place where `utils` have to live

Thanks for feedback!

**Comment by RingerJK** [comment]
@Guardiola31337  `NullUtils` are used into `utils` module internally, hold 2 copies of this class is necessary because `base` and `utils` don't connect to each other.  Because of this we cannot have only 1 copy of `NullUtils`. 

Regarding your term that we shouldn't bring whole module to use only one method. Actually we made it with `utils` module before, when we need `NullUtils` cross-modules; only one aim of `utils ` module is share common approaches so it looks that all modules should depend on `util`

**Comment by Guardiola31337** [comment]
@RingerJK 
>`NullUtils` are used into `utils` module internally, hold 2 copies of this class is necessary because `base` and `utils` don't connect to each other.  Because of this we cannot have only 1 copy of `NullUtils`.

Good point. What about having a bit of duplication but decreasing it to the max? I mean, what about only defining the `ifNonNull` functions that we use on each module and remove the rest? We can always add new ones if needed in the future. What do you think?

> Regarding your term that we shouldn't bring whole module to use only one method. Actually we made it with `utils` module before, when we need `NullUtils` cross-modules; only one aim of `utils ` module is share common approaches so it looks that all modules should depend on `util`

Previously we've discussed the tradeoffs on a case by case basis. I'd rather continuing doing that instead of adding `utils` dependency across all modules, as sometimes it's preferable introducing a bit of duplication at the cost of not bringing a whole library.

**Comment by RingerJK** [comment]
@Guardiola31337 
>Good point. What about having a bit of duplication but decreasing it to the max? I mean, what about only defining the ifNonNull functions that we use on each module and remove the rest? We can always add new ones if needed in the future. What do you think?

What is intention of this approach? `ifNonNull` is small piece of sugar, not a huge library logic. It might need in any time in different branches 🤔 

>Previously we've discussed the tradeoffs on a case by case basis. I'd rather continuing doing that instead of adding utils dependency across all modules, as sometimes it's preferable introducing a bit of duplication at the cost of not bringing a whole library.

What pros of non-including `utils` module into all modules? I'm sure that utils should, instead copy-paste:
- `utils` already into almost all modules, except `base` and `libnavigator`; so even if customer exclude almost all modules `utils` still be in the library. 
-  `util` is lightweight and doesn't depend on any module, compile time increased really small.

because of ^^ `utils` doesn't increase size of `aar` or compile time, but avoid copy-paste code. IMO `utils` should do included into all modules

## I think we should resolve the main issue here
 I'd like to show how I see this dependencies:
`onboard`/`offboard`/`notifation` and other modules consume `utils` and use anything from it
`base` cannot consume `utls` so it has own `internal` `utls` and doesn't share between modules. Moreover, share utils is not a goal of `base`. `base` is about contracts between modules, we shouldn't mix goals of `base` and give additional functional to it, let's keep `base` as pure interface-provider module and `utils` as utils-provider.

File: libnavigator/build.gradle [hunk 1]
```diff
@@ -31,6 +31,7 @@ android {
 
 dependencies {
     implementation(project(':libnavigation-base'))
+    implementation project(':libnavigation-util')
```
[the remark this answers is not in the document]
Comment (Guardiola31337) [reply libnavigator/build.gradle:34]: What about 👇?

```diff
diff --git a/libnavigation-base/src/main/java/com/mapbox/navigation/base/extensions/NullUtils.kt b/libnavigation-base/src/main/java/com/mapbox/navigation/base/extensions/NullUtils.kt
deleted file mode 100644
index c4cbadaa2..000000000
--- a/libnavigation-base/src/main/java/com/mapbox/navigation/base/extensions/NullUtils.kt
+++ /dev/null
@@ -1,10 +0,0 @@
-@file:JvmName("NullUtils")
-
-package com.mapbox.navigation.base.extensions
-
-internal inline fun <R1, T> ifNonNull(r1: R1?, func: (R1) -> T): T? =
-    if (r1 != null) {
-        func(r1)
-    } else {
-        null
-    }
diff --git a/libnavigation-base/src/main/java/com/mapbox/navigation/base/route/internal/RouteUrl.kt b/libnavigation-base/src/main/java/com/mapbox/navigation/base/route/internal/RouteUrl.kt
index 0d0314229..e45d6dce3 100644
--- a/libnavigation-base/src/main/java/com/mapbox/navigation/base/route/internal/RouteUrl.kt
+++ b/libnavigation-base/src/main/java/com/mapbox/navigation/base/route/internal/RouteUrl.kt
@@ -2,7 +2,7 @@ package com.mapbox.navigation.base.route.internal
 
 import android.net.Uri
 import com.mapbox.geojson.Point
-import com.mapbox.navigation.base.extensions.ifNonNull
+import com.mapbox.navigation.base.utils.internal.ifNonNull
 
 class RouteUrl(
     val accessToken: String,
diff --git a/libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/internal/NullUtils.kt b/libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/internal/NullUtils.kt
new file mode 100644
index 000000000..05b4d2ae1
--- /dev/null
+++ b/libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/internal/NullUtils.kt
@@ -0,0 +1,10 @@
+@file:JvmName("NullUtils")
+
+package com.mapbox.navigation.base.utils.internal
+
+inline fun <R1, T> ifNonNull(r1: R1?, func: (R1) -> T): T? =
+    if (r1 != null) {
+        func(r1)
+    } else {
+        null
+    }
diff --git a/libnavigator/build.gradle b/libnavigator/build.gradle
index b9bd50748..27f0259e6 100644
--- a/libnavigator/build.gradle
+++ b/libnavigator/build.gradle
@@ -31,7 +31,6 @@ android {
 
 dependencies {
     implementation(project(':libnavigation-base'))
-    implementation project(':libnavigation-util')
 
     // Navigator
     api dependenciesList.mapboxNavigator
diff --git a/libnavigator/src/main/java/com/mapbox/navigation/navigator/MapboxNativeNavigatorImpl.kt b/libnavigator/src/main/java/com/mapbox/navigation/navigator/MapboxNativeNavigatorImpl.kt
index 1841010d7..53ff0f1be 100644
--- a/libnavigator/src/main/java/com/mapbox/navigation/navigator/MapboxNativeNavigatorImpl.kt
+++ b/libnavigator/src/main/java/com/mapbox/navigation/navigator/MapboxNativeNavigatorImpl.kt
@@ -17,7 +17,7 @@ import com.mapbox.navigation.base.trip.model.RouteLegProgress
 import com.mapbox.navigation.base.trip.model.RouteProgress
 import com.mapbox.navigation.base.trip.model.RouteProgressState
 import com.mapbox.navigation.base.trip.model.RouteStepProgress
-import com.mapbox.navigation.utils.extensions.ifNonNull
+import com.mapbox.navigation.base.utils.internal.ifNonNull
 import com.mapbox.navigator.BannerComponent
 import com.mapbox.navigator.BannerInstruction
 import com.mapbox.navigator.BannerSection
```

File: libnavigator/src/main/java/com/mapbox/navigation/navigator/MapboxNativeNavigatorImpl.kt [hunk 2]
```diff
@@ -13,11 +13,11 @@ import com.mapbox.geojson.Geometry
 import com.mapbox.geojson.Point
 import com.mapbox.geojson.gson.GeometryGeoJson
 import com.mapbox.geojson.utils.PolylineUtils
-import com.mapbox.navigation.base.extensions.ifNonNull
 import com.mapbox.navigation.base.trip.model.RouteLegProgress
 import com.mapbox.navigation.base.trip.model.RouteProgress
 import com.mapbox.navigation.base.trip.model.RouteProgressState
 import com.mapbox.navigation.base.trip.model.RouteStepProgress
+import com.mapbox.navigation.utils.extensions.ifNonNull
```
Comment (Guardiola31337) [inline libnavigator/src/main/java/com/mapbox/navigation/navigator/MapboxNativeNavigatorImpl.kt:20]: See comment above

**Comment by RingerJK** [comment]
@Guardiola31337 great suggestion, but it ignores arguments from discussion above. Take a look at https://github.com/mapbox/mapbox-navigation-android/pull/2701/commits/38dc61ee21925dc3fcb8c81e413139727fe20324 😉

File: libnavigator/src/main/java/com/mapbox/navigation/navigator/utils/NullUtils.kt [hunk 3]
```diff
@@ -0,0 +1,10 @@
+@file:JvmName("NullUtils")
+
+package com.mapbox.navigation.navigator.utils
+
+internal inline fun <R1, T> ifNonNull(r1: R1?, func: (R1) -> T): T? =
```
Comment (Guardiola31337) [inline libnavigator/src/main/java/com/mapbox/navigation/navigator/utils/NullUtils.kt:5]: I'd rather making `base` `NullItils` "public" (under an `internal` package) as suggested https://github.com/mapbox/mapbox-navigation-android/pull/2701 but :+1:

**Pull request merged**